classes.py

```diff
@@ -0,0 +1,544 @@
+        for i in range(1, len(rows)):
+            cells = rows[i].find_all("td", recursive=False)
+            id_opponent = cells[id_cell_num].get_text().strip()
+            name_opponent = cells[name_cell_num].get_text().strip()
+            result_opponent = cells[result_cell_num].get_text().strip()
+            if name_opponent not in ['not paired', 'bye']:
+                self.add_opponent(int(id_opponent), name_opponent, result_opponent)
+
+    def add_opponent(self, sno, name, result):
+        if result[-1] != "K":
+            if result[-1] == "½":
+                res = '0.5'
+            else:
+                res = result[-1]
+            self.opponents.append([sno, name, float(res)])
+
+    def keep_current_rating(self):
+        # self.new_rating = self.this_rating = self.last_rating
+        self.new_rating = self.last_rating
+        self.new_total_games = self.last_total_games
+        self.new_sum_oppon_ratings = self.last_sum_oppon_ratings
+        self.new_pts_against_oppon = self.last_pts_against_oppon
+
+    def calculate_new_rating(self, is_fexerj_tournament):
+
+        invalid_opponents = []
+        for k, tp_oppon in self.opponents.items():
+            if tp_oppon[0].is_unrated:
+                if self.is_unrated or (tp_oppon[0].new_rating is None) or tp_oppon[0].new_rating == 0:
+                    invalid_opponents.append(k)
+        for i in invalid_opponents:
+            del self.opponents[i]
+
+        self.this_games = len(self.opponents)
+
+        if self.this_games == 0 or (self.is_unrated and self.this_pts_against_oppon == 0):
+            self.keep_current_rating()
+            return
+
+        self.this_sum_oppon_ratings = 0
+        self.this_pts_against_oppon = 0
+        for snr_opp, oppon in self.opponents.items():
+            if oppon[0].is_unrated:  # unrated x unrated will never happen
+                self.this_sum_oppon_ratings += oppon[0].new_rating
+            elif oppon[0].is_temp and not (self.is_unrated or self.is_temp):  # self established x oppon temporary
+                self.this_sum_oppon_ratings += oppon[0].new_rating
+            else:
+                self.this_sum_oppon_ratings += oppon[0].last_rating
+            self.this_pts_against_oppon += oppon[1]
+
+        if self.is_unrated and self.this_pts_against_oppon == 0:
+            self.keep_current_rating()
+            return
+
+        self.last_k = self.get_current_k()
+        self.this_avg_oppon_rating = self.this_sum_oppon_ratings / self.this_games
+        rating_diff = self.this_avg_oppon_rating - self.last_rating
+        self.this_expected_points = self.this_games / (1.0 + 10.0 ** (rating_diff / 400.0))
+        self.this_points_above_expected = (self.this_pts_against_oppon - self.this_expected_points)
+
+        self.new_total_games = self.last_total_games + self.this_games
+
+        calc_rule = self.get_calculation_rule(is_fexerj_tournament)
+        if calc_rule == "TEMPORARY":
```

Comment:
We should also use enums for the calculation rule, and ideally have separate functions for each one of them (instead of having the entire logic in this function in each if-else branch)